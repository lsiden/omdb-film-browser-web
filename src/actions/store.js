import { createStore, applyMiddleware, compose } from "redux"
import thunk from "redux-thunk"

import Actions from "actions/types"

const initialState = {
  view: Actions.VIEW_FILM_LIST,
}

export function reduce(state = initialState, action) {
  const { type, data } = action
  switch (type) {
    case Actions.VIEW_FILM_LIST:
    case Actions.VIEW_FILM_DETAIL:
    case Actions.UPDATE_FILMS:
    case Actions.UPDATE_FILM_DETAILS:
    case Actions.UPDATE_IS_FETCHING:
      return {
        ...state,
        ...data,
      }
    case Actions.APPEND_FILMS:
      return {
        ...state,
        ...data,
        films: [...state.films, ...data.films],
      }
    default:
      return state
  }
}

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose
const store = createStore(reduce, composeEnhancers(applyMiddleware(thunk)))

export default store
