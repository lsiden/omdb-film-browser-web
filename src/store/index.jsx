import { createStore, applyMiddleware, compose } from 'redux'
import thunk from 'redux-thunk'

import ActionType from './const'
import reduce from './reducers'

// Action factories

export const setQuery = (query='') => ({
  type: ActionType.SET_QUERY,
  data: { query },
})

export const replaceFilms = (query='', films=[], totalResults=0) => ({
  type: ActionType.REPLACE_FILMS,
  data: { films, query, totalResults, pageNum: 1 },
})

export const appendFilms = (pageNum=1, films=[]) => ({
  type: ActionType.APPEND_FILMS,
  data: { films, pageNum },
})

export const updateFilmDetails = (filmDetails) => ({
  type: ActionType.UPDATE_FILM_DETAILS,
  data: { filmDetails },
})

export const setFetching = (isFetching=false) => ({
  type: ActionType.UPDATE_IS_FETCHING,
  data: { isFetching }
})

/// //////////////
// State Accessors

export function getFilms(state) {
  return Object.values(state.films || {})
}

/// /////////////
// createStore()

const composeEnhancers = (
  window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__
  && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({ trace: true, traceLimit: 25 })
) || compose

export default createStore(reduce, composeEnhancers(applyMiddleware(thunk)))
