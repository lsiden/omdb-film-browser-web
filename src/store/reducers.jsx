import ActionTypes from 'store/const'

/*
 * I discovered that OMDB search can return some duplicate films.
 * This will cause React to throw a warning when it discovers a duplicate key.
 * So we have to guard against duplicate items.
 */
function reduceUniqueFilms(res, film) {
  res[film.imdbID] = film
  return res
}

const reduceNewFilms = (state, data) => ({
  ...state,
  films: (data.films || []).reduce(reduceUniqueFilms, {}),
  totalResults: parseInt(data.totalResults || '0', 10),
  pageNum: 1,
})

const reduceMoreFilms = (state, data) => ({
  ...state,
  films: (data.films || []).reduce(reduceUniqueFilms, { ...state.films || {} }),
  pageNum: data.pageNum,
})

export default function (state={}, action) {
  const { type, data } = action
  switch (type) {
    case ActionTypes.UPDATE_FILM_DETAILS:
    case ActionTypes.UPDATE_IS_FETCHING:
    case ActionTypes.SET_QUERY:
      return {
        ...state,
        ...data,
      }
    case ActionTypes.REPLACE_FILMS:
      return reduceNewFilms(state, data)
    case ActionTypes.APPEND_FILMS:
      return reduceMoreFilms(state, data)
    default:
      return state
  }
}
