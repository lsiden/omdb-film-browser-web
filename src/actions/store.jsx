import { createStore, applyMiddleware, compose } from 'redux'
import thunk from 'redux-thunk'

import Actions from 'actions/types'
import { updateNewFilms, appendFilms } from 'actions/films'

const initialState = {
  view: Actions.VIEW_FILM_LIST,
  isFetching: false,
  films: [],
  totalResults: 0,
  query: '',
}

export function reduce(state = initialState, action) {
  const { type, data } = action
  switch (type) {
    case Actions.UPDATE_FILM_DETAILS:
    case Actions.UPDATE_IS_FETCHING:
    case Actions.SET_QUERY:
      return {
        ...state,
        ...data,
      }
    case Actions.UPDATE_FILMS:
      return updateNewFilms(state, data)
    case Actions.APPEND_FILMS:
      return appendFilms(state, data)
    default:
      return state
  }
}

const composeEnhancers = (
    window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__
    && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({ trace: true, traceLimit: 25 })
  ) || compose
const store = createStore(reduce, composeEnhancers(applyMiddleware(thunk)))

export default store
