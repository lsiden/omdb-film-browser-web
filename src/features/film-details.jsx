import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import Spinner from 'respin'

import CloseButton from 'components/close-button'
import NavButton from 'components/nav-button'
import { scrollToTop } from 'components/scroll'
import { fetchFilmDetails } from 'actions/remote'
import { headerStyle as bannerHeaderStyle } from 'style'
import { ESC_KEY, BANNER_TITLE, FETCH_TIMEOUT } from 'constants'

const titleStyle = {
  marginBottom: '5pt',
}
const detailsStyle = {
  marginTop: 20,
  fontSize: '14pt',
}
const topButtonStyle = {
  marginBottom: 20,
}
const spinnerWrapperStyle = {
  height: '100vh',
  minHeight: '100vh',
  display: 'flex',
  justifyContent: 'center',
}

const itemExists = (item) => item && item !== 'N/A'
const imdbUrl = (imdbID) => `https://www.imdb.com/title/${imdbID}`

const titleBannerStyle = {
  ...bannerHeaderStyle,
  display: 'flex',
  justifyContent: 'space-between',
}

const TitleBanner = ({ title }) => (
  <header style={titleBannerStyle}>
    <div style={{ lineHeight: 1 }}>
      <h1 style={{ fontSize: 24 }}>{BANNER_TITLE}</h1>
      <br />
      <h1 style={{ ...titleStyle, marginTop: 0 }}>{title}</h1>
    </div>
    <CloseButton buttonStyle={{ color: 'white' }} />
  </header>
)
TitleBanner.propTypes = {
  title: PropTypes.string.isRequired,
}

export class FilmDetails_ extends React.Component {
  static keyDownListener(ev) {
    if (ev.keyCode === ESC_KEY) {
      window.location.href = '/'
    }
  }

  constructor(props) {
    const { imdbID } = props
    super(props)
    this.imdbUrl = `https://www.imdb.com/title/${imdbID}`
  }

  componentDidMount() {
    document.addEventListener('keydown', FilmDetails_.keyDownListener)
    setTimeout(() => {
      if (!this.props.filmDetails) {
        const { dispatchFetchFilmDetails, imdbID } = this.props
        dispatchFetchFilmDetails(imdbID)
      }
    }, FETCH_TIMEOUT)
  }

  componentWillUnmount() {
    document.removeEventListener('keydown', FilmDetails_.keyDownListener)
  }

  renderTitle() {
    const { filmDetails } = this.props
    return <TitleBanner title={filmDetails.Title} />
  }

  renderDetails() {
    const { filmDetails } = this.props
    const FilmDetail = ({ name, label }) => (
      <li>{label || name}: {filmDetails[name]}</li>
    )

    return (
      <ul style={detailsStyle}>
        <FilmDetail name="Year" />
        <FilmDetail name="Director" />
        <FilmDetail name="Writer" />
        <FilmDetail name="Actors" label="Cast" />
        <FilmDetail name="Language" />
        {itemExists(filmDetails.Awards) && <FilmDetail name="Awards" />}
        <FilmDetail name="Runtime" label="Length" />
        <FilmDetail name="imdbRating" label="IMDB Rating" />
        <FilmDetail name="BoxOffice" />
        {
          itemExists(filmDetails.Website)
          && <FilmDetail key="Website" label="Official Website" />
        }
        <li>
          <a href={imdbUrl(filmDetails.imdbID)}>IMDB page</a>
        </li>
      </ul>
    )
  }

  static renderScrollToTopButton() {
    return (
      <NavButton
        style={topButtonStyle}
        onClick={scrollToTop}
        title="Scroll to top of page"
      >
        Top
      </NavButton>
    )
  }

  render() {
    const { filmDetails } = this.props

    if (!filmDetails) {
      return (
        <div style={spinnerWrapperStyle}>
          <Spinner size={64} />
        </div>
      )
    } else {
      return (
        <div>
          {this.renderTitle()}
          <div style={{ marginLeft: 20 }}>
            <img
              src={filmDetails.Poster}
              alt="poster"
              style={{ marginTop: 10 }}
            />
            {this.renderDetails()}
            {FilmDetails_.renderScrollToTopButton()}
          </div>
        </div>
      )
    }
    return null
  }
}

FilmDetails_.propTypes = {
  imdbID: PropTypes.string.isRequired,
  filmDetails: PropTypes.object,
  dispatchFetchFilmDetails: PropTypes.func.isRequired,
}
FilmDetails_.defaultProps = {
  filmDetails: null,
}

export default connect(
  (state) => ({
    filmDetails: state.filmDetails,
  }),
  (dispatch) => ({
    dispatchFetchFilmDetails: (imdbID) => dispatch(fetchFilmDetails(imdbID)),
  }),
)(FilmDetails_)
