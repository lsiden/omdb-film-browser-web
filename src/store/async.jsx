// Remote Actions

import toastr from 'toastr'
// import promiseFinally from 'promise.prototype.finally'

import store from 'store'
import { FETCH_TIMEOUT, OMDB_API_URL, OMDB_API_KEY } from 'omdb_constants'
import {
  replaceFilms,
  appendFilms,
  updateFilmDetails,
  setFetching,
} from '.'

const queryFetchUri = (query) => `${OMDB_API_URL}?apikey=${OMDB_API_KEY}&type=movie&s=${query}`
const pageFetchUri = (query, page) => queryFetchUri(query) + `&page=${page}`
const filmFetchUri = (id) => `${OMDB_API_URL}?apikey=${OMDB_API_KEY}&type=movie&i=${id}&plot=full`

function toJson(res) {
  try {
    return res.json()
  } catch (e) {
    return Promise.reject(e)
  }
}

const onCatch = (e) => {
  console.error(e)
  toastr.error(e, 'An error occured', {
    preventDuplicates: true,
  })
}

// Returns Promise that resolves to fetched result if successful
const promiseFetchOrTimeout = (dispatch, uri, millisec=FETCH_TIMEOUT) => {
  let timeout
  const timeoutPromise = new Promise((_, reject) => {
    timeout = setTimeout(() => {
      reject(
        new Error(
          'Unable to fetch data. Try a different title or try again later.',
        ),
      )
    }, millisec)
  })
  dispatch(setFetching(true))
  return Promise.race([fetch(uri), timeoutPromise])
    .then(toJson)
    .catch(onCatch)
    .finally(() => {
      clearTimeout(timeout)
      dispatch(setFetching(false))
    })
}

export const promiseQueryResults = (query) => (dispatch) => promiseFetchOrTimeout(
  dispatch, queryFetchUri(query)
).then((res={ Search: [], totalResults: 0 }) => dispatch(replaceFilms(
  query, res.Search, res.totalResults
)))

export const promiseQueryPageFetch = () => (dispatch) => {
  const { query='', pageNum=1 } = store.getState()
  const nextPageNum = pageNum + 1
  return promiseFetchOrTimeout(dispatch, pageFetchUri(query, nextPageNum))
    .then((res={ Search: [], totalResults: 0 }) => {
      dispatch(appendFilms(nextPageNum, res.Search))
    })
}

export const promiseFilmDetails = (id) => (dispatch) => promiseFetchOrTimeout(
  dispatch,
  filmFetchUri(id),
).then((res) => dispatch(updateFilmDetails(res)))
