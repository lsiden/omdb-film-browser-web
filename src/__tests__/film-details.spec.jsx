import React from 'react'
import TestRenderer from 'react-test-renderer'
import { Provider } from 'react-redux'

import { FilmDetails_ } from 'film-details'
import store from 'actions/store'

const filmDetails = require('./film-details.json')
const { imdbID } = filmDetails

it('renders details when present', () => {
  const params = {
    imdbID,
    filmDetails,
    dispatchFetchFilmDetails: jest.fn(),
  }
  expect(
    TestRenderer.create(
      <FilmDetails_ { ...params } />
    ),
  ).toMatchSnapshot()
})

it('renders a spinner when details not present', () => {
  const params = {
    imdbID,
    dispatchFetchFilmDetails: jest.fn(),
  }
  expect(
    TestRenderer.create(
      <FilmDetails_ { ...params } />
    ),
  ).toMatchSnapshot()
})
