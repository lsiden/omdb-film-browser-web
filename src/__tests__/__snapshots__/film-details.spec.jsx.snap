// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a spinner when details not present 1`] = `
<div
  style={
    Object {
      "display": "flex",
      "height": "100vh",
      "justifyContent": "center",
      "minHeight": "100vh",
    }
  }
>
  <svg
    fill="currentcolor"
    height={64}
    style={
      Object {
        "display": "inline-flex",
        "margin": 0,
      }
    }
    viewBox="0 0 32 32"
    width={64}
  >
    <path
      d="M14 0 H18 V8 H14z"
      opacity={0.1}
      transform="rotate(0 16 16)"
    >
      <animate
        attributeName="opacity"
        begin="0ms"
        dur="1000ms"
        from={1}
        repeatCount="indefinite"
        to={0.1}
      />
    </path>
    <path
      d="M14 0 H18 V8 H14z"
      opacity={0.1}
      transform="rotate(45 16 16)"
    >
      <animate
        attributeName="opacity"
        begin="125ms"
        dur="1000ms"
        from={1}
        repeatCount="indefinite"
        to={0.1}
      />
    </path>
    <path
      d="M14 0 H18 V8 H14z"
      opacity={0.1}
      transform="rotate(90 16 16)"
    >
      <animate
        attributeName="opacity"
        begin="250ms"
        dur="1000ms"
        from={1}
        repeatCount="indefinite"
        to={0.1}
      />
    </path>
    <path
      d="M14 0 H18 V8 H14z"
      opacity={0.1}
      transform="rotate(135 16 16)"
    >
      <animate
        attributeName="opacity"
        begin="375ms"
        dur="1000ms"
        from={1}
        repeatCount="indefinite"
        to={0.1}
      />
    </path>
    <path
      d="M14 0 H18 V8 H14z"
      opacity={0.1}
      transform="rotate(180 16 16)"
    >
      <animate
        attributeName="opacity"
        begin="500ms"
        dur="1000ms"
        from={1}
        repeatCount="indefinite"
        to={0.1}
      />
    </path>
    <path
      d="M14 0 H18 V8 H14z"
      opacity={0.1}
      transform="rotate(225 16 16)"
    >
      <animate
        attributeName="opacity"
        begin="625ms"
        dur="1000ms"
        from={1}
        repeatCount="indefinite"
        to={0.1}
      />
    </path>
    <path
      d="M14 0 H18 V8 H14z"
      opacity={0.1}
      transform="rotate(270 16 16)"
    >
      <animate
        attributeName="opacity"
        begin="750ms"
        dur="1000ms"
        from={1}
        repeatCount="indefinite"
        to={0.1}
      />
    </path>
    <path
      d="M14 0 H18 V8 H14z"
      opacity={0.1}
      transform="rotate(315 16 16)"
    >
      <animate
        attributeName="opacity"
        begin="875ms"
        dur="1000ms"
        from={1}
        repeatCount="indefinite"
        to={0.1}
      />
    </path>
  </svg>
</div>
`;

exports[`renders details when present 1`] = `
<div>
  <header
    style={
      Object {
        "background": "linear-gradient(to right, #000040, #3b00ff)",
        "color": "white",
        "display": "flex",
        "justifyContent": "space-between",
        "padding": "10px",
      }
    }
  >
    <div
      style={
        Object {
          "lineHeight": 1,
        }
      }
    >
      <h1
        style={
          Object {
            "fontSize": 24,
          }
        }
      >
        Open Movie Database Browser
      </h1>
      <br />
      <h1
        style={
          Object {
            "marginBottom": "5pt",
            "marginTop": 0,
          }
        }
      >
        The Battle of Algiers
      </h1>
    </div>
    <button
      className="btn btn-link"
      onClick={[Function]}
      style={
        Object {
          "color": "white",
          "fontSize": 20,
          "margin": 20,
        }
      }
      type="button"
    >
      âœ•
    </button>
  </header>
  <div
    style={
      Object {
        "marginLeft": 20,
      }
    }
  >
    <img
      alt="poster"
      src="https://images-na.ssl-images-amazon.com/images/M/MV5BYzliNzI5NzgtNDZmNy00NGU4LTg2OTQtMDEzZTRiNDU3ZTI1XkEyXkFqcGdeQXVyMDUyOTUyNQ@@._V1_SX300.jpg"
      style={
        Object {
          "marginTop": 10,
        }
      }
    />
    <ul
      style={
        Object {
          "fontSize": "14pt",
          "marginTop": 20,
        }
      }
    >
      <li>
        Year
        : 
        1966
      </li>
      <li>
        Director
        : 
        Gillo Pontecorvo
      </li>
      <li>
        Writer
        : 
        Franco Solinas, Franco Solinas (story), Gillo Pontecorvo (story)
      </li>
      <li>
        Cast
        : 
        Jean Martin, Yacef Saadi, Brahim Hadjadj, Tommaso Neri
      </li>
      <li>
        Language
        : 
        French, Arabic, English
      </li>
      <li>
        Awards
        : 
        Nominated for 3 Oscars. Another 9 wins & 5 nominations.
      </li>
      <li>
        Length
        : 
        121 min
      </li>
      <li>
        IMDB Rating
        : 
        8.1
      </li>
      <li>
        BoxOffice
        : 
        $55,908
      </li>
      <li>
        Official Website
        : 
      </li>
      <li>
        <a
          href="https://www.imdb.com/title/tt0058946"
        >
          IMDB page
        </a>
      </li>
    </ul>
    <button
      className="btn btn-link"
      onClick={[Function]}
      style={
        Object {
          "marginBottom": 20,
        }
      }
      type="button"
    >
      Top
    </button>
  </div>
</div>
`;
